src/data/matrix/kronecker.lean

```diff
@@ -0,0 +1,317 @@
+/-
+Copyright (c) 2021 Filippo A. E. Nuccio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo A. E. Nuccio, Eric Wieser
+-/
+
+import data.matrix.basic
+import linear_algebra.tensor_product
+import ring_theory.tensor_product
+
+/-!
+# Kronecker product of matrices
+
+This defines the [Kronecker product](https://en.wikipedia.org/wiki/Kronecker_product).
+
+## Main definitions
+
+* `matrix.kronecker_map`: A generalization of the Kronecker product, defined such that
+  `kronecker_map f A B (i₁, i₂) (j₁, j₂) = f (A i₁ j₁) (B i₁ j₂)`.
+* `matrix.kronecker_map_linear`: when `f` is bilinear, so is `kronecker_map f`.
+
+## Specializations
+
+* `matrix.kronecker`: An alias of `kronecker_map (*)`. Prefer using the notation.
+* `matrix.kronecker_linear`: `matrix.kronecker` is bilinear
+
+* `matrix.kronecker_tmul`: An alias of `kronecker_map (⊗ₜ)`. Prefer using the notation.
+* `matrix.kronecker_tmul_linear`: `matrix.tmul_kronecker` is bilinear
+
+## Notations
+
+These require `open_locale kronecker`:
+
+* `A ⊗ₖ B` for `kronecker_map (*) A B`
+* `A ⊗ₖₜ B` and `A ⊗ₖₜ[R] B` for `kronecker_map (⊗ₜ) A B`
+
+-/
+
+namespace matrix
+
+open_locale matrix
+
+variables {R α α' β β' γ γ' : Type*}
+variables {l m n p : Type*} [fintype l] [fintype m] [fintype n] [fintype p]
+variables {l' m' n' p' : Type*} [fintype l'] [fintype m'] [fintype n'] [fintype p']
+
+section kronecker_map
```

Comment:
Wouldn't it be better to have a `namespace kronecker_prod` and to drop `kronecker` from every name in all the lemmas below?

Reply:
Normally namespace are reserved for types and not functions, as dot notation never works on the latter; we don't have `add.zero` or `tmul.add` for instance. That's not to say it would be a bad idea; just that the value is marginal, and there isn't much precedent for naming in that way.

Reply:
Oh, I see: I certainly take this back as you have more insight into these matters.